libp2p/network/network_test.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/pokt-network/pocket/shared/crypto"
+)
+
+const testIP6ServiceURL = "2a00:1450:4005:802::2004:8080"
+
+func TestLibp2pNetwork_AddPeerToAddrBook(t *testing.T) {
+	p2pNet := newTestLibp2pNetwork(t)
+	peerstore := p2pNet.host.Peerstore()
+
+	// NB: assert initial state
+	require.Len(t, p2pNet.addrBookMap, 1)
+
+	var (
+		existingAddrStr string
+		existingPeer    *types.NetworkPeer
+	)
+	for addr, peer := range p2pNet.addrBookMap {
+		existingAddrStr = addr
+		existingPeer = peer
+	}
+	require.NotEmpty(t, existingAddrStr)
+	require.NotNil(t, existingPeer)
+
+	existingPeerInfo, err := Libp2pAddrInfoFromPeer(existingPeer)
+	require.NoError(t, err)
+
+	existingPeerstoreAddrs := peerstore.Addrs(existingPeerInfo.ID)
+	require.Len(t, existingPeerstoreAddrs, 1)
+
+	existingPeerMultiaddr, err := Libp2pMultiaddrFromServiceURL(existingPeer.ServiceURL)
+	require.NoError(t, err)
+	require.Equal(t, existingPeerstoreAddrs[0].String(), existingPeerMultiaddr.String())
+
+	newPublicKey, err := crypto.GeneratePublicKey()
+	newPoktAddr := newPublicKey.Address()
+	require.NoError(t, err)
+
+	newPeer := &types.NetworkPeer{
+		PublicKey:  newPublicKey,
+		Address:    newPoktAddr,
+		ServiceURL: testIP6ServiceURL,
+	}
+	newPeerInfo, err := Libp2pAddrInfoFromPeer(newPeer)
+	require.NoError(t, err)
+	newPeerMultiaddr := newPeerInfo.Addrs[0]
+
+	// NB: add to address book
+	err = p2pNet.AddPeerToAddrBook(newPeer)
+	require.NoError(t, err)
+
+	require.Len(t, p2pNet.addrBookMap, 2)
+	require.Equal(t, p2pNet.addrBookMap[existingAddrStr], existingPeer)
+	require.Equal(t, p2pNet.addrBookMap[newPeer.Address.String()], newPeer)
+
+	existingPeerstoreAddrs = peerstore.Addrs(existingPeerInfo.ID)
+	newPeerstoreAddrs := peerstore.Addrs(newPeerInfo.ID)
+
+	require.Len(t, existingPeerstoreAddrs, 1)
+	require.Len(t, newPeerstoreAddrs, 1)
+	require.Equal(t, newPeerstoreAddrs[0].String(), newPeerMultiaddr.String())
+}
+
+func TestLibp2pNetwork_RemovePeerToAddrBook(t *testing.T) {
+	p2pNet := newTestLibp2pNetwork(t)
+	peerstore := p2pNet.host.Peerstore()
+
+	// NB: assert initial state
+	require.Len(t, p2pNet.addrBookMap, 1)
+
+	var (
+		existingAddrStr string
+		existingPeer    *types.NetworkPeer
+	)
+	for addr, peer := range p2pNet.addrBookMap {
+		existingAddrStr = addr
+		existingPeer = peer
+	}
+	require.NotEmpty(t, existingAddrStr)
+	require.NotNil(t, existingPeer)
+
+	existingPeerInfo, err := Libp2pAddrInfoFromPeer(existingPeer)
+	require.NoError(t, err)
+
+	existingPeerstoreAddrs := peerstore.Addrs(existingPeerInfo.ID)
+	require.Len(t, existingPeerstoreAddrs, 1)
+
+	existingPeerMultiaddr, err := Libp2pMultiaddrFromServiceURL(existingPeer.ServiceURL)
+	require.NoError(t, err)
+	require.Equal(t, existingPeerstoreAddrs[0].String(), existingPeerMultiaddr.String())
+
+	err = p2pNet.RemovePeerFromAddrBook(existingPeer)
+	require.NoError(t, err)
+
+	require.Len(t, p2pNet.addrBookMap, 0)
+
+	// NB: peerstore implementations seem to only remove peer keys and
```

Comment:
This is definitely pretty weird behaviour...

Reply:
I think it has to do with the TTL concept that's associated with addresses in the libp2p peerstore.

Reply:
Nice that we have the comment.